sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala

```diff
@@ -148,7 +148,7 @@ object OptimizeSkewedJoin extends CustomShuffleReaderRule {
   /*
    * This method aim to optimize the skewed join with the following steps:
```

Comment:
immature thoughts on this optimization:

`OptimizeSkewedJoin` try to make sizes of shuffle partitions more even. 

For bhj, should we change the *optimization goal of build side* to making sure build side fit in memory to avoid OOM or fallback to smj.

Is it possible that the build side is skewed according current criterions, but the shuffle partitions all fit in memory?

Reply:
In current implementaion, we have already supported optimize skew build side with inner join. For other join type we cann't optimize it due to the semantics.

IMO It's better to consider OOM at shj itself instead of fallback to smj which might make things more complicated. Actually, we might change smj to shj at `reOptimize` if we disable the `preferSortMerge`, so it's confused to change join strategy again.

Reply:
Sorry my comment was misleading, the `fallback to smj` there means https://github.com/apache/spark/pull/32210

My thought is that we may take potential OOM into account here,
let build side A inner join stream side B, for a build partition A_0:

- if A_0 is skewed but is less than a OOM threshold, we may not split it;
- if A_0 is not skewed but greater than that threshold, we may still need to split it;

Reply:
My hunch is when the build side can be potentially OOM-ed, it should already be considered as skewed. So after AQE skew handling, some of potentially OOM-ed build side (inner join only) can be avoided.

However, for queries with other join types, queries not having shuffle before join, and queries with run-time hash map being significantly larger than partition size, we should have run-time fallback mechanism in shuffled hash join itself. This PR and #32210 should be good to have and orthogonal to each other.